tests/ui/update_banner.js

```diff
@@ -0,0 +1,96 @@
+ /*
+     Test for the update banner states, using the mozApps mock to pretend
+     Marketplace app is installed/has an update available/etc.
+ */
+var helpers = require('../lib/helpers');
+
+var bannerSelector = '#marketplace-update-banner';
+var downloadButtonSelector = bannerSelector + ' .download-button';
+
+
+casper.test.begin('Test manifest_url', {
+    test: function(test) {
+        helpers.startCasper();
+        helpers.waitForPageLoaded(function() {
+            test.assertTruthy(casper.evaluate(function() {
+                return window.require('core/settings').manifest_url;
+            }), 'Check that manifest_url is defined in settings');
+        });
+        helpers.done(test);
+    }
+});
+
+
+casper.test.begin('Test show banner and click download button', {
+    test: function(test) {
+        helpers.startCasper();
+        helpers.waitForPageLoaded(function() {
+            casper.evaluate(function() {
+                var manifest = window.require('core/settings').manifest_url;
+                window.require('core/settings').is_iframed_app = true;
+                window.navigator.mozApps.install(manifest);
+                window.navigator.mozApps._setDownloadAvailable(manifest, true);
+                window.require('update_banner').showIfNeeded();
+            });
+
+            helpers.assertWaitForSelector(test, bannerSelector, function() {
+                test.assertSelectorExists(downloadButtonSelector);
+                casper.click(downloadButtonSelector);
+            });
+
+            casper.waitWhileSelector(downloadButtonSelector, function() {
+                test.assertSelectorHasText(bannerSelector,
+                    'The next time you start the Firefox Marketplace app');
+                test.assertNotExists(downloadButtonSelector);
+            });
+
+
+        });
+        helpers.done(test);
+    }
+});
+
+
+casper.test.begin('Test banner is not shown if no download is available', {
+    test: function(test) {
+        helpers.startCasper();
+        helpers.waitForPageLoaded(function() {
+            casper.evaluate(function() {
+                var manifest = window.require('core/settings').manifest_url;
+                window.require('core/settings').is_iframed_app = true;
+                window.navigator.mozApps.install(manifest);
+                window.require('update_banner').showIfNeeded();
+            });
+
+            // Wait a little before checking if the banner exists, but not long
+            // since our mozApps mock makes the whole thing synchronous.
+            casper.wait(250, function() {
```

Comment:
I wasn't sure ; since the selector isn't present in the first place (no banner initially), I think `casper.waitWhileSelector` would immediately resolve. I wanted to wait a little to make sure that even after all the async checks are made, there is still no update banner added